File: app/src/main/java/com/example/pictureme/views/home/HomeFragment.java
package com.example.pictureme.views.home;

import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.pictureme.R;
import com.example.pictureme.models.Picme;
import com.example.pictureme.viewmodel.PicmeListViewModel;
import com.example.pictureme.views.home.adapters.HomeSectionsAdapter;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class HomeFragment extends Fragment {

    private RecyclerView recyclerView;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_home, container, false);

        PicmeListViewModel picmeListViewModel = new ViewModelProvider(getActivity()).get(PicmeListViewModel.class);

        recyclerView = view.findViewById(R.id.recycler_view_sections);

        List<Picme>[] sectionPicmes = new List[3];
        sectionPicmes[0] = new ArrayList<>();
        sectionPicmes[1] = new ArrayList<>();
        sectionPicmes[2] = new ArrayList<>();

        HomeSectionsAdapter mAdapter = new HomeSectionsAdapter(
                new String[]{"PicMe's with friends", "Food PicMe's", "PicMe's from long ago"},
                sectionPicmes
        );

        picmeListViewModel.getUserPicmesLiveData().observe(getViewLifecycleOwner(), picmes -> {
            System.out.println("HELLO1");
            sectionPicmes[0] = picmes;
            sectionPicmes[1] = picmes;
            sectionPicmes[2] = picmes;
            mAdapter.notifyDataSetChanged();
        });

        recyclerView.setAdapter(mAdapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(view.getContext()));


        return view;
    }
}

File: app/src/main/java/com/example/pictureme/views/home/HomeFragment.kt
package com.example.pictureme.views.home

import android.os.Build
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.RequiresApi
import androidx.core.view.isGone
import androidx.core.view.isVisible
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.pictureme.R
import com.example.pictureme.views.home.adapters.HomeAdapter
import com.example.pictureme.data.models.Feeling
import com.example.pictureme.data.models.Friendship
import com.example.pictureme.data.models.Picme
import com.example.pictureme.databinding.FragmentHomeBinding
import com.example.pictureme.utils.FilterPicmes
import com.example.pictureme.viewmodels.PicmeDetailsViewModel
import com.example.pictureme.viewmodels.PicmeViewModel
import com.example.pictureme.viewmodels.UserViewModel
import dagger.hilt.android.AndroidEntryPoint
import java.util.*

@AndroidEntryPoint
class HomeFragment : Fragment() {

    private var _binding: FragmentHomeBinding? = null
    private val binding get() = _binding!!

    private val userViewModel by activityViewModels<UserViewModel>()
    private val picmeViewModel by activityViewModels<PicmeViewModel>()
    private val picmeDetailsViewModelViewModel by activityViewModels<PicmeDetailsViewModel>()

    @RequiresApi(Build.VERSION_CODES.O)
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentHomeBinding.inflate(inflater, container, false)

        // Welcoming message based on day time
        setWelcomingMessage()

        // PicMe Cards Setup
        setPicMeCards()

        return (binding.root)
    }

    // Set welcoming message based on day time
    private fun setWelcomingMessage() {
        val rightNow = Calendar.getInstance()
        val currentHour: Int = rightNow.get(Calendar.HOUR_OF_DAY) // 24 hour format
        userViewModel.userLiveData.observe(viewLifecycleOwner) { user ->
            val fullName = user.fullName
            val firstName = fullName!!.split(" ")[0]
            if (currentHour < 12)
                binding.fragmentHomeTvWelcome.text = "Good Morning, $firstName"
            else if (currentHour < 18)
                binding.fragmentHomeTvWelcome.text = "Good Afternoon, $firstName"
            else
                binding.fragmentHomeTvWelcome.text = "Good Evening, $firstName"
        }

    }

    // Setup PicMe Cards
    @RequiresApi(Build.VERSION_CODES.O)
    private fun setPicMeCards() {
        var feelings: List<Feeling> = emptyList()
        var picmes: List<Picme> = emptyList()

        val rvHome = binding.fragmentHomeRv
        val rvsCategory = arrayListOf<ParentModelClass>()

        val filteredPicmes = FilterPicmes.getFilteredPicmes(picmes, feelings)
        for (filter in filteredPicmes) {
            val rv = ParentModelClass(filter.first, filter.second)
            rvsCategory.add(rv)
        }

        val adapter = HomeAdapter(rvsCategory, picmeDetailsViewModelViewModel, requireContext())
        rvHome.adapter = adapter
        rvHome.layoutManager =
            LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false)

        picmeViewModel.foodFeelingsLiveData.observe(viewLifecycleOwner) { response ->
            feelings = response
        }

        // If data changes, update
        picmeViewModel.picmesLiveData.observe(viewLifecycleOwner) { response ->
            //println("DATA CHANGED: ${response[response.size - 1]}")
            val rvsCategory = arrayListOf<ParentModelClass>()
            val filteredPicmes = FilterPicmes.getFilteredPicmes(response, feelings)

            if (filteredPicmes.isEmpty()) {
                binding.textNoPicmes.visibility = View.VISIBLE
                binding.fragmentHomeIv.visibility = View.VISIBLE
            } else {
                binding.textNoPicmes.visibility = View.GONE
                binding.fragmentHomeIv.visibility = View.GONE
            }

            for (filter in filteredPicmes) {
                val rv = ParentModelClass(filter.first, filter.second)
                rvsCategory.add(rv)
            }
            adapter.setList(rvsCategory)

            binding.rvLoadingBar.isGone = true
            binding.fragmentHomeRv.visibility = View.VISIBLE
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

}